api/sys_raft.go

```diff
@@ -0,0 +1,91 @@
+package api
+
+import (
+	"context"
+	"io"
+)
+
+type RaftJoinResponse struct {
+	Joined bool `json:"joined"`
+}
+
+type RaftJoinRequest struct {
+	LeaderAddr string `json:"leader_api_addr"`
+	CACert     string `json:"ca_cert":`
+	Retry      bool   `json:"retry"`
+}
+
+func (c *Sys) RaftJoin(opts *RaftJoinRequest) (*RaftJoinResponse, error) {
+	r := c.c.NewRequest("POST", "/v1/sys/storage/raft/join")
+
+	if err := r.SetJSONBody(opts); err != nil {
+		return nil, err
+	}
+
+	ctx, cancelFunc := context.WithCancel(context.Background())
+	defer cancelFunc()
+	resp, err := c.c.RawRequestWithContext(ctx, r)
+	if err != nil {
+		return nil, err
+	}
+	defer resp.Body.Close()
+
+	var result RaftJoinResponse
+	err = resp.DecodeJSON(&result)
+	return &result, err
+}
+
+func (c *Sys) RaftSnapshot(snapWriter io.Writer) error {
+	r := c.c.NewRequest("GET", "/v1/sys/storage/raft/snapshot")
+
+	ctx, cancelFunc := context.WithCancel(context.Background())
+	defer cancelFunc()
+	resp, err := c.c.RawRequestWithContext(ctx, r)
+	if err != nil {
+		return err
+	}
+
+	io.Copy(snapWriter, resp.Body)
```

Comment:
I think you should be checking the error returned by copy

Reply:
Also we shouldn't be using RawRequestWithContext here since it parses the response body here: https://github.com/hashicorp/vault/blob/170521481d58961761bfd6877221b005cf0f4c7e/api/response.go#L37-L40

Reply:
Fixed.